cache/src/test/java/io/envoyproxy/controlplane/cache/SimpleCacheTest.java

```diff
@@ -159,6 +184,67 @@ public void successfullyWatchAllResourceTypesWithSetBeforeWatchWithRequestVersio
     }
   }
 
+  @Test
+  public void successfullyWatchAllResourceTypesWithSetBeforeWatchWithSameRequestnVersionNewResourceHints() {
+    SimpleCache<String> cache = new SimpleCache<>(new SingleNodeGroup());
+
+    cache.setSnapshot(SingleNodeGroup.GROUP, MULTIPLE_RESOURCES_SNAPSHOT2);
+
+    // Set a watch for the current snapshot with the same version but with resource hints present
+    // in the snapshot that the watch creator does not currently know about.
+    Map<String, Watch> watches = Resources.TYPE_URLS.stream()
+        .collect(Collectors.toMap(
+            typeUrl -> typeUrl,
+            typeUrl -> cache.createWatch(
+                ADS,
+                DiscoveryRequest.newBuilder()
+                    .setNode(Node.getDefaultInstance())
+                    .setTypeUrl(typeUrl)
+                    .setVersionInfo(MULTIPLE_RESOURCES_SNAPSHOT2.version(typeUrl))
+                    .addAllResourceNames(NAMES_MULTIPLE_RESOURCES_SNAPSHOT.get(typeUrl))
```

Comment:
It took me a little while to parse what was going on here. Maybe it would be more intuitive if this line used `MULTIPLE_RESOURCES_SNAPSHOT2.resources(typeUrl).keySet()` or if `SNAPSHOT2.resources(typeUrl).keySet()` below was changed to `NAMES.get(typeUrl)`? In hindsight, maybe that `NAMES` collection doesn't even need to exist?

It may also be helpful to add some comments here and below to explicitly call out the requested resources versus the existing known resources.